vignettes/algorithm.Rmd

```diff
@@ -0,0 +1,341 @@
+                  length.out = nlambda))
+head(lambda)
+```
+
+`lambda.min.ratio` and `nlambda` are both setable through the API, but 
+are given here at their defaults (per this example). Note that we return the
+lambda path on the original scale of the respone, $\mathbf{y}$, which means 
+that the user sees
+
+```{r}
+lambda_out <- lambda*y_sd
+head(lambda_out)
+```
+
+Naturally, whatever value the user provides in the argument `lambda` to
+`sgdnet()` will be scaled accordingly.
+
+### Step size
+
+The step size, $\gamma$, in SAGA is constant throughout the algorithm.
+For non-strongly convex objectives it is set to $1/(3L)$, where $L$ is the 
+*Lipschitz* constant. Since we are picking a single sample at a time,
+we approximate this constant by the largest sample-wise squared norm
+of the feature matrix. For strongly convex objectives,
+the step size is set to
+
+$$ \frac{1}{3(\mu n + L)},$$
+
+where $\mu$ is the level of strong convexity and $n$ the number of samples.
+For our purposes, $\mu$, is simply the regularization strength of
+the $\ell_2$-penalty: $\alpha\lambda$. This has the effect of 
+letting the step size adapt to the type of regularization penalty used.
+
+In our example, where we have the elastic net penalty, we will have the
+following step sizes (one for each $\lambda$). For the sake of illustration,
+we will also include the case of ridge and lasso penalties to see the 
+differences in step size.
+
+```{r, fig.width = 7, fig.cap = "Step sizes along the regularization path."}
+reg <- sapply(c(0, 0.5, 1), function(alpha_i) {
+  lambda_2 <- (1 - alpha_i)*lambda/n # l2 penalty
+  L <- max(rowSums(x^2)) + lambda_2
+  n <- nrow(x)
+  mu <- 2*lambda_2
+  step_size <- 1 / (2*(L + pmin(mu*n, L)))
+})
+reg <- as.data.frame(reg)
+colnames(reg) <- c("ridge", "elastic_net", "lasso")
+
+library(lattice)
+xyplot(ridge + elastic_net + lasso ~ seq_len(nlambda), 
+       data = reg, type = "l",
+       xlab = NULL,
+       ylab = expression(gamma),
+       auto.key = list(lines = TRUE, points = FALSE, space = "right"))
+
+```
+
+### Algorithm
+
+We will now look at the implementation (in pseudocode) of the actual SAGA
+algorithm as it is realized in **sgdnet**.
+
+```{r, eval = FALSE}
+y           response vector of size n
+X           feature matrix of size n*p, each sample X[j] is stored columnwise
+beta        vector of coefficients of size p
+g_avg       gradient average
+l1          amount of L1-regularization
+l2          amount of L2-regularization
+max_iter    maximum number of outer iterations
+gamma       the step size
+lag         a vector of size dnenoting the last time each feature was updated
+lag_scaling a series of geometric sums to use with LaggedUpdate()
+
+for i in 1:max_iter
+  for k in 1:n_samples
+    X[j] <- draw a sample randomly from X
+    
+    nz <- Nonzeros(X[j]) # nonzero indices of X[j]
+    
+    # Perform just-in-time updates of coefficients X is sparse
+    X[j] <- LaggedUpdate(j,
```

Comment:
For simplicity, I might focus on the dense (no lagged update) case first and then describe the lagged updates near the end.